// eventbrite-node/lib/eventbrite.js
'use strict';

var request = require('request'),
  Q = require('q'),
  url = require('url'),
  assert = require('assert'),
  utils = require('./utils'),
  auth = require('../../../config/auth.js'),
  noop = function() {
  };

var CLIENTKEY = auth.eventbrite.clientID;
var CLIENTSECRET = auth.eventbrite.clientSecret;
var OAUTH_URL = auth.eventbrite.oauthURL;
var AUTHORIZATION_URL = auth.eventbrite.authorizationURL;
var API_URL = auth.eventbrite.apiURL;
var ACCESS_TOKEN = '';

module.exports = Eventbrite;    
                        
/**
 *
 * Authorization flow:
 *
 * 1. set clientKey & clientSecret
 * 2. get OAuth url & redirect user
 * 3. extract code from response & exchange it to token
 * 4. talk with eventbrite
 *
 *
 * @constructor
 * @param {string} clientKey Identifies your app during the OAuth handshake. Not secret.
 * @param {string} clientSecret Identifies your app during a server-side handshake. Secret.
 */
function Eventbrite(clientKey, clientSecret) {

  if (!(this instanceof Eventbrite)) {
    return new Eventbrite(clientKey, clientSecret);
  }
 // this.accessToken = null;
 // this.clientKey = clientKey || null;
 //  this.clientSecret = clientSecret || null;
}

Q.denodeify(Eventbrite.get);

Eventbrite.prototype = {

  /**
   * Set access token
   * @param {string} accessToken
   * @returns {Eventbrite}
   */
  setAccessToken: function(accessToken) {
    this.accessToken = accessToken;
    return this;
  },

  /**
   * Return url which should be used to redirect user
   *
   * @returns {string}
   */
  getOAuthUrl: function() {
    var params = {
        response_type: 'code',
        client_id: CLIENTKEY
      },
      urlObj = url.parse(OAUTH_URL);

    urlObj.query = params;
    return url.format(urlObj);
  },

  /**
   * Make authorization - exchange code to access token
   *
   * @param {string} code
   */
  authorize: function(code, callback) {
   callback = callback || noop;
   var formData = {
          code: code,
          client_secret: CLIENTSECRET,
          client_id: CLIENTKEY,
          grant_type: 'authorization_code'
        };
      
    request.post({
        url: AUTHORIZATION_URL,
        headers: {
          'Content-type': 'application/x-www-form-urlencoded'
        },
        form: formData
      },
      function(err, response, body) {
        if (err) {
          body = utils.toJSON(body);
          console.log('error is :' , err);
          console.log('body is :' , body);
          console.log('body error :' , body.error);
          callback(null, { error: body.error });
          //return callback(err);
          //http://localhost:5504/networks/geteventbritetokenstep2/?error=access_denied
        }
        else {
          body = utils.toJSON(body);
          ACCESS_TOKEN = body.access_token;
          callback(null, { access_token: body.access_token });
        }
      }.bind(this));
  },


  /*
  get: function(path,  params, callback ) {
    if(callback === undefined) {
      callback = params;
      params = {};
    }

    callback = callback || noop;

    if(path[0] === '/') {
      path = path.substr(1);
    }
    var queryParams = {
      url: url.resolve(API_URL, path),
      qs: params,
      headers: {
        Authorization: 'Bearer ' + 'P3KTZDWBSBEH7BSBAGS2'
      }
    };

    request.get(queryParams,
      function(err, response, body) {
        if (err) {
          return callback(response);
        }
        body = utils.toJSON(body);
        callback(null, body );
      }.bind(this));
  },

  post: function(path, params, callback) {
   // assert(this.accessToken, 'The value of access token is not set');

    if(callback ===undefined) {
      callback = params;
      params = {};
    }

    callback = callback || noop;
    if(path[0] === '/') {
      path = path.substr(1);
    }
    request.post({
        url: url.resolve(API_URL, path),
        json: params,
        headers: {
          Authorization: 'Bearer ' + ACCESS_TOKEN
        }
      },
      function(err, response, body) {
        if (err) {
          return callback(response);
        }
        //callback(null, utils.toJSON(body));
        return body;
      });
  }

  */
};

Eventbrite.prototype.authorize = Q.denodeify(Eventbrite.prototype.authorize);
Eventbrite.prototype.get = Q.denodeify(Eventbrite.prototype.get);
Eventbrite.prototype.post = Q.denodeify(Eventbrite.prototype.post);
